scripts/tests/py/metadata.py

```diff
@@ -0,0 +1,225 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Union
+
+import yaml
+
+
+@dataclass
+class Metadata:
+    py_script_path: Optional[str] = None
+    run: Optional[str] = None
+    app: Optional[str] = None
+    factoryreset: bool = False
+    comissioning_method: Optional[str] = None
+    discriminator: Optional[int] = None
+    kvs: Optional[str] = None
+    storage_path: Optional[str] = None
+    on_network_commission: Optional[str] = None
+    passcode: Optional[int] = None
+    endpoint: Optional[str] = None
+    manual_code: Optional[str] = None
+    PICS: Optional[str] = None
+    tests: Optional[str] = None
+
+    def copy_from_dict(self, attr_dict: Dict[str, str]) -> None:
+        """
+        Sets the value of the attributes from a dictionary.
+
+        Attributes:
+
+        attr_dict:
+          Dictionary that stores attributes value that should
+          be transferred to this class.
+        """
+
+        if "app" in attr_dict:
+            self.app = attr_dict["app"]
+
+        if "run" in attr_dict:
+            self.run = attr_dict["run"]
+
+        if "discriminator" in attr_dict:
+            self.discriminator = attr_dict["discriminator"]
+
+        if "passcode" in attr_dict:
+            self.passcode = attr_dict["passcode"]
+
+        if "py_script_path" in attr_dict:
+            self.py_script_path = attr_dict["py_script_path"]
+
+        # TODO - set other attributes as well
+
+
+class MetadataReader:
+    """
+    A class to parse run arguments from the test scripts and 
+    resolve them to environment specific values.
+    """
+
+    def __init__(self, env_yaml_file_path: str):
+        """
+        Reads the YAML file and Constructs the environment object
+
+        Parameters:
+
+        env_yaml_file_path:
+          Path to the environment file that contains the YAML configuration.
+        """
+        with open(env_yaml_file_path) as stream:
+            self.env = yaml.safe_load(stream)
+
+    def __resolve_env_vals__(self, metadata_dict: Dict[str, Union[str, bool]]) -> None:
+        """
+        Resolves the argument defined in the test script to environment values.
+        For example, if a test script defines "all_clusters" as the value for app
+        name, we will check the environment configuration to see what raw value is
+        assocaited with the "all_cluster" variable and set the value for "app" option
+        to this raw value.
+
+        Parameter:
+
+        metadata_dict:
+          Dictionary where each key represent a particular argument and its value represent
+          the value for that argument defined in the test script.
+        """
+
+        for run_arg, run_arg_val in metadata_dict.items():
+
+            if not type(run_arg_val) == str or run_arg == "run":
+                metadata_dict[run_arg] = run_arg_val
```

Comment:
why do we have this assignment? is this not a noop given that you get the arg and value ... and this just sets the arg and value again?

Reply:
A comment would be helpful here, to explain why non-strings and runs are skipped.